### PR TITLE
Highlight works fine with the font size plugin

### LINKED ISSUE #17: Highlight does not work well with font size
image

Looks like a matter of conversion priorities.

image

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Fix: Highlight plugin will render the highlighted text properly when it's combined with the `FontSize` plugin. Closes #17.

### CODE REVIEW

**Comment by pomek** [comment]
I forgot to do some internal checks. I'm giving R- now.

**Comment by pomek** [comment]
Should be fine now. Let's wait for @coveralls report.

File: src/highlightediting.js [hunk 1]
```diff
@@ -107,11 +115,36 @@ function _buildDefinition( options ) {
...
 	for ( const option of options ) {
 		definition.model.values.push( option.model );
-		definition.view[ option.model ] = {
-			name: 'mark',
-			classes: option.class
+
+		definition.view[ option.model ] = ( modelAttributeValue, viewWriter ) => {
+			const attributes = { class: option.class };
+
+			// Highlight element has to have higher priority than other view elements because it must sticks directly to the text.
```
Comment (Reinmar) [inline src/highlightediting.js]: must stick

**Comment by Reinmar** [comment]
Shouldn't we increase the priority of the font-size feature instead of patching the highlight? What about font-size + any other feature which changes text background?

I think that this is a very similar case than the link feature. It got itself the highest priority. 

cc @pjasiun

**Comment by jodator** [comment]
> Shouldn't we increase the priority of the font-size feature

Did you mean to lower font-size priority? Here we had higher priority for Highlight so to make it work in Font feature I assume that it have to have lower priority then Highlight (or other features) to properly change wrapped elements - ie be the latest one to convert.

**Comment by pjasiun** [comment]
> Shouldn't we increase the priority of the font-size feature instead of patching the highlight? What about font-size + any other feature which changes text background?
> 
> I think that this is a very similar case than the link feature. It got itself the highest priority.
> 
> cc @pjasiun

TBH, I don't know the context, but if a problem can be solved by view attribute elements priorities we should do it, instead of hacking.

**Comment by Reinmar** [comment]
The latest one to convert will be the link feature.

Here, we have to decide whether the highlight feature requires a proper treatment. Or whether it's the font-size feature. 

My point was that I'm unsure whether highlight has to be the innermost tag. Perhaps it's the font-size which should be close-to-outermost. There's, in fact, a difference in these two solutions. This PR will make the highlight tags rendered as close to the text as possible, which also involves moving it inside `<strong>`, `<i>`, etc. Does that make sense?

**Comment by Reinmar** [comment]
> TBH, I don't know the context, but if a problem can be solved by view attribute elements priorities we should do it, instead of hacking.

That's exactly what's proposed ;) The question is – should we increase highlight's priority or decrease font-size's priority?

**Comment by jodator** [comment]
I'm thinking that what causes problems here is font-size changing feature + background coloring feature combo.

As for now I can see:
- font-size + highlight
- font-size + link-highlight (the link highlight has the highest priority of them all)

In future we might have font-background plugin that will have to behave similarly to highlight. So probably making the font-size outer most (lower its priority) will be more future proof. WDYT?

**Comment by pjasiun** [comment]
As you said:

> this is a very similar case than the link feature. It got itself the highest priority.

So in this case highlight should also get a higher priority. It should work well with all features which change formatting.

**Comment by Reinmar** [comment]
> In future we might have font-background plugin that will have to behave similarly to highlight. So probably making the font-size outer most (lower its priority) will be more future proof. WDYT?

👍

**Comment by Reinmar** [comment]
We agreed to first add the ability to specify the view attribute element priority in two-way converters: https://github.com/ckeditor/ckeditor5-engine/issues/1408.

**Comment by pomek** [comment]
After introducing https://github.com/ckeditor/ckeditor5-engine/issues/1408, this PR is no longer valid. See https://github.com/ckeditor/ckeditor5-font/pull/16.

**Pull request closed without merging**